### PR TITLE
feat: compare models side by side for cost and speed

### PR DESCRIPTION
**Pull request by brunobuddy**
### 💭 Why
PR #1787 (the Benchmark page) went stale and conflicting after 227 commits landed on main. This brings it current, resolves the conflicts, and restores full coverage on the new code.

### ✨ What changed
- Resolve additive merge conflicts: `database.module`, e2e helpers, frontend deps, `api` barrel, regenerated lockfile.
- Rename the benchmark migration `1785000000000` → `1788000000000` (two existing migrations already use `1785000000000`).
- Fix a latent flaky `pickDefaults` test: native detection is now slug-aware, so `openai/gpt-4o-mini` on the openai provider beats a `gemini-` imposter instead of a coin-flip shuffle.
- Drop dead `providerIdForModel` from `benchmark-store`.
- Add tests for the previously-untested benchmark UI (column, empty state, model picker, icons, markdown, summary table, page, store) → 100% line coverage on every new file.

### 👤 For users
New Benchmark page in the sidebar: send one prompt to several models at once and compare cost, output tokens, and latency. Past runs are saved per agent with a history drawer; optional custom request headers; responses render as sanitized markdown.

### 🔧 For operators
One new migration (`benchmark_runs`, `benchmark_columns`). TypeORM runs it automatically on boot. No env vars or config changes.

### 📝 Notes
- Supersedes #1787 (same feature, conflicts resolved). Close #1787 once this merges.
- `benchmark` is promoted to a first-class message tier, so benchmark calls show in the Messages list under their own filter.
- Verified locally: 4762 backend unit, 164 e2e, 3073 frontend tests pass; backend + frontend tsc clean.

---
## Summary by cubic
Adds a Benchmark page to compare multiple models side by side for cost, tokens, and latency. Includes saved per‑agent history with replay, sanitized markdown output with hardened links, and optional custom request headers; failed columns are now saved to history too.

- **New Features**
  - New Benchmark page with parallel runs, per-column cancel, and a summary table highlighting cheapest and fastest.
  - History: persists runs (successes and failures) per user+agent, shows Today/Yesterday/Earlier, and can replay a past run into the view.
  - Security: sanitizes request headers on input and whitelists safe response headers; renders markdown via `dompurify` + `marked` and forces rel="noopener noreferrer" on target links.
  - Backend: adds `BenchmarkModule` with run + history endpoints; saves columns and metrics; writes messages with routing_tier=`benchmark`.
  - Message Log: adds a Tier filter (uses `ALL_TIERS` / `TIER_LABELS_ALL` from `manifest-shared`) so benchmark calls are easy to find.

- **Migration**
  - Adds tables: `benchmark_runs` and `benchmark_columns` (TypeORM migration runs on boot).
  - No config or env changes.
  - Frontend deps: adds `dompurify`, `marked`, and `@types/dompurify`.

<sup>Written for commit a83e14df4ff171d57eec76cc3f5fd90e27a1cf96. Summary will update on new commits. <a href="https://cubic.dev/pr/mnfst/manifest/pull/1934?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

File: packages/backend/src/benchmark/benchmark-payload.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      }
+    }
+  }
+  // OpenAI Responses API: { input: "..." } or { input: [{role, content}, ...] }
+  const input = body['input'];
+  if (typeof input === 'string') return input;
+  if (Array.isArray(input)) {
+    for (let i = input.length - 1; i >= 0; i--) {
+      const item = input[i] as { role?: unknown; content?: unknown } | null;
+      if (item && item.role === 'user' && typeof item.content === 'string') return item.content;
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/benchmark/benchmark-payload.ts]: Fixed in a83e14df4

File: packages/backend/src/benchmark/benchmark-payload.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+function derivePromptFromRawBody(body: Record<string, unknown>): string {
+  // OpenAI chat-completions / Anthropic messages: { messages: [{role, content}, ...] }
+  const messages = body['messages'];
+  if (Array.isArray(messages)) {
+    for (let i = messages.length - 1; i >= 0; i--) {
+      const m = messages[i] as { role?: unknown; content?: unknown } | null;
+      if (!m || m.role !== 'user') continue;
+      if (typeof m.content === 'string') return m.content;
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/benchmark/benchmark-payload.ts]: Fixed in a83e14df4

File: packages/backend/src/benchmark/dto/run-benchmark.dto.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  @IsNotEmpty()
+  provider!: string;
+
+  @IsOptional()
+  @IsIn(AUTH_TYPES)
+  authType?: 'api_key' | 'subscription' | 'local';
+
+  @IsOptional()
+  @IsArray()
+  @ArrayMinSize(1)
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/benchmark/dto/run-benchmark.dto.ts]: Fixed in a83e14df4

File: packages/backend/src/benchmark/benchmark.service.ts [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        status: 'error',
+        content: null,
+        headers,
+        errorMessage: errorSummary,
+        metrics: null,
+      });
+      throw new BadGatewayException(errorSummary);
+    }
+
+    const openaiBody = this.normalizeResponseBody(bodyText, dto.model, formatFlags);
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/benchmark/benchmark.service.ts]: Fixed in a83e14df4

File: packages/backend/src/benchmark/benchmark.service.ts [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+      });
+      rawResponse = fwd.response;
+      formatFlags = {
+        isGoogle: fwd.isGoogle,
+        isAnthropic: fwd.isAnthropic,
+        isChatGpt: fwd.isChatGpt,
+      };
+    } catch (err) {
+      const message = err instanceof Error ? err.message : String(err);
+      throw new BadGatewayException(`Provider request failed: ${message}`);
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/backend/src/benchmark/benchmark.service.ts]: Fixed in a83e14df4

File: packages/frontend/src/components/benchmark/MarkdownContent.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+interface Props {
+  text: string;
+  class?: string;
+}
+
+const MarkdownContent: Component<Props> = (props) => {
+  const html = createMemo(() => {
+    const parsed = marked.parse(props.text ?? '') as string;
+    return DOMPurify.sanitize(parsed, {
+      ALLOWED_ATTR: ['href', 'title', 'class', 'target', 'rel'],
```
[the remark this answers is not in the document]
Comment (brunobuddy) [reply packages/frontend/src/components/benchmark/MarkdownContent.tsx:53]: Fixed in a83e14df4
